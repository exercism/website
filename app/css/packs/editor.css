@import "../minimal_tailwind.css";
@import "../ui-kit/minimal.css";

@import "../pages/editor";
@import "../modals/editor-hints";
@import "../components/chatgpt";
