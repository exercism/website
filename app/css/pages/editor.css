@import "../styles";

@keyframes progress0To100 {
    from {
        width: 0%;
    }
    to {
        width: 100%;
    }
}

#page-editor {
    @apply fixed flex flex-col;
    width: 100%;
    height: 100%;

    & .header {
        @apply bg-backgroundColorA;
        @apply border-b-1 border-borderColor6;
        @apply flex-shrink-0 flex-grow-0;

        @apply flex items-center;
        @apply py-6 px-32;

        & .close-btn,
        & .options {
            flex-basis: 0;
            flex-grow: 1;
        }

        & .close-btn {
            @apply flex;
            @apply font-mono font-medium text-textColor6 text-14;
            & .c-icon {
                height: 16px;
                width: 16px;
                @apply mr-16;
                filter: var(--textColor6-filter);
            }
        }
        & .title {
            @apply flex-grow flex justify-center;
            @apply font-mono text-15 font-medium;

            & .track {
                @apply text-textColor6;
            }

            & .divider {
                @apply text-textColor6;
                @apply px-16;
            }

            & .exercise {
                @apply text-textColor2 font-medium;
            }
        }
        & .options {
            @apply flex items-center justify-end;
        }

        & .hints-btn,
        & .keyboard-shortcuts-btn,
        & .settings-btn,
        & .more-btn {
            height: 36px;
            @apply flex justify-center items-center;
        }

        & .hints-btn {
            @apply px-6;
            & .c-icon {
                filter: var(--orange-filter);
                height: 24px;
                width: 24px;
            }
        }

        & .keyboard-shortcuts-btn {
            @apply px-6;
            @apply text-textColor6;
            & .c-icon {
                height: 24px;
                width: 24px;
                filter: var(--textColor6-filter);
            }
        }

        & .hints-btn,
        & .keyboard-shortcuts-btn,
        & .settings-btn {
            @apply mr-24;
        }

        & .settings-btn,
        & .more-btn {
            @apply text-textColor6;
            @apply px-10;
            & .c-icon {
                height: 16px;
                width: 16px;
                filter: var(--textColor6-filter);
            }
        }

        & .settings-dialog {
            @apply bg-backgroundColorA;
            @apply flex flex-col shadow-lg;
            @apply py-24 px-32 text-left;
            @apply z-dropdown;

            & .setting {
                @apply flex items-center justify-between;
                @apply mb-16;
                &:last-child {
                    @apply mb-0;
                }

                & .name {
                    @apply text-16 font-medium text-darkGray;
                    @apply mr-56;
                    width: 110px;
                }

                & .options {
                    @apply bg-borderLight rounded-100;
                    @apply flex items-center flex-grow-0;
                    box-shadow: 0px 4px 16px rgba(0, 0, 0, 0.2);

                    & label {
                        @apply block relative;
                        & .label {
                            @apply px-12 py-6;
                            @apply cursor-pointer;
                            @apply border-1 border-borderLight rounded-100;
                            @apply text-textColor6 font-mono font-bold leading-140;
                        }

                        /* Keep it here so it's discoverable by touch
                         * for accessibility purposes */
                        & input {
                            @apply absolute;
                            top: 0;
                            left: 0;
                            opacity: 0;
                            &:checked + .label {
                                @apply bg-lightBlue;
                                @apply text-textColor9;
                            }

                            /* It would be nice to DRY this up with application.css
                             * but its tricky with the adjecent element. */
                            &:focus {
                                & + .label {
                                    outline: 5px auto -webkit-focus-ring-color;
                                }
                                &:not(.focus-visible) + .label {
                                    outline: none;
                                }
                            }
                        }
                    }
                }
            }
        }
        & .actions-dialog {
            min-width: 380px;
        }
    }
    /* This appears in various places in the UI */
    /* This whole thing is to get round issues with
     * disabled buttons not firing mouse events */
    & .run-tests-btn {
        @apply relative;

        & .disabled-wrapper {
            @apply hidden absolute;
            top: 0;
            left: 0;
            right: 0;
            bottom: 0;
        }
        & button[disabled] + .disabled-wrapper {
            @apply block;
            cursor: not-allowed;
        }
    }

    & .c-split-pane {
        & .--split-lhs {
            @apply flex flex-col;
            @apply bg-backgroundColorB;
            @apply overflow-hidden;
            width: calc(100% - 33%);

            & .tabs {
                @apply flex;
                & .c-tab {
                    @apply mr-0;
                }
            }

            & .c-iteration-pane {
                overflow: hidden;
            }
            & .--tab-panel {
                @apply flex-grow;
                @apply overflow-hidden;
            }
            & .editor {
                @apply flex-grow;
                @apply bg-backgroundColorA;
                overflow-y: auto;
                height: 100%;

                & .ace_content {
                    @apply font-medium;
                }

                /* These bits handle things resizing correctly */
                @apply flex;
                & .ace_editor {
                    @apply flex-grow;
                    @apply w-100;
                }
            }

            /*@apply border-t-1 border-borderColor6;*/
            & footer.lhs-footer {
                @apply flex-shrink-0 flex-grow-0;
                @apply flex items-center justify-end px-32 py-24;
                @apply bg-backgroundColorB;
                @apply relative;
                box-shadow: inset 0 24px 24px -24px var(--backgroundColorA);

                & button {
                    @apply ml-20;
                }
                & .run-tests-btn {
                    & button {
                        @apply ml-0;
                    }
                }
            }
        }
        & .--split-rhs {
            @apply bg-backgroundColorA;
            @apply flex flex-col;
            width: 33%;

            /* Put a shadow over the LHS */
            @apply shadow-xsZ1 z-shadow;

            & .tabs {
                @apply flex flex-shrink-0 flex-grow-0;
                @apply py-16 px-24;

                & .c-tab {
                    height: 36px;
                    @apply mr-12;

                    & .c-icon {
                        height: 16px;
                        width: 16px;
                    }
                }
            }

            & > .--tab-panel {
                overflow-y: scroll;
                height: 100%;
            }

            & section.instructions-pane {
                @apply pb-16 px-24;

                & > .c-textual-content {
                    & * + h3 {
                        margin-top: 0;
                    } /* Override the default */
                    & :not(h2) + h3 {
                        @apply mt-28;
                    }
                    & .content {
                        @apply mt-0;
                    }
                }
                & .instructions {
                    @apply pt-16;
                    @apply border-t-1 border-borderColor6;
                }
                & .c-details.task {
                    @apply border-t-2 border-lightGray;
                    @apply mt-24 pt-16;

                    & > .--summary {
                        @apply mb-10;
                        & > .--summary-inner {
                            @apply flex items-center;

                            & .task-marker {
                                @apply mr-20;
                                @apply py-4 px-8;
                                @apply bg-lightGold text-unnamed13 rounded-3;
                                @apply font-mono font-bold text-12 leading-120;
                                @apply flex-shrink-0 self-start;
                                margin-top: 2px;
                            }
                            & .summary-title {
                                @apply flex-grow;
                                @apply text-18 font-semibold leading-140;
                            }

                            & .--closed-icon,
                            & .--open-icon {
                                height: 24px;
                                width: 24px;

                                @apply flex flex-shrink-0 items-center justify-center;
                                @apply border-gray border-1 rounded-circle;
                                @apply text-darkGray;
                                @apply ml-16;

                                & .c-icon {
                                    height: 11px;
                                    width: 11px;
                                }
                            }
                        }
                    }
                }
            }
            & .--tab-panel.tests {
                &:not([hidden]) {
                    @apply flex;
                }
                & code {
                    @apply pb-24;
                    height: 100%;
                }
            }
            & section.results {
                & .running {
                    @apply px-24 py-24;
                    @apply flex flex-col items-center;
                    & .c-icon {
                        height: 24px;
                        width: 24px;
                        animation: spin 3s linear infinite;
                        filter: var(--textColor6-filter);
                        @apply mb-20;
                    }
                    & .progress {
                        @apply w-100 rounded-100;
                        @apply bg-veryLightBlue;
                        @apply flex items-center;
                        @apply mb-12;
                        height: 6px;

                        & .bar {
                            height: 6px;
                            min-width: 6px;
                            background: linear-gradient(
                                to bottom,
                                #2200ff 0%,
                                #9e00ff 100%
                            );

                            @apply rounded-100;

                            animation: progress0To100 normal forwards;
                            animation-timing-function: cubic-bezier(
                                0.82,
                                0.27,
                                0.05,
                                0.92
                            );
                            animation-iteration-count: 1;
                            animation-duration: 3s;
                        }
                    }
                    & p {
                        @apply text-14 leading-170 font-mono font-medium;
                        @apply text-textColor6;
                        @apply mb-16;
                        & strong {
                            @apply font-semibold;
                            @apply text-textColor2;
                        }
                    }
                }
            }
            & section.run-tests-prompt {
                @apply flex flex-col items-center;
                @apply pt-32 pb-8 px-8 mb-24;

                & .c-icon {
                    height: 48px;
                    width: 48px;
                    @apply mb-16;
                }

                & h2 {
                    @apply flex items-center;
                    @apply mb-12;
                    &,
                    & button {
                        @apply text-h6 text-textColor6;
                    }
                    & button {
                        @apply py-0 px-0 mr-4;
                        & :first-child {
                            @apply mr-4;
                        }
                        & .--kb {
                            padding-top: 3px;
                            padding-bottom: 2px;
                        }
                    }
                }
                & p {
                    @apply text-15 leading-160 text-textColor6;
                    @apply text-center;
                    max-width: 450px;
                }
            }
        }
    }
}

#page-editor {
    & .cm-editor {
        width: 100%;

        &.cm-focused,
        & .focus-visible {
            /* TODO: Find a way to set outline for a11y only */
            outline: none;
        }
        & .cm-panels-top {
            @apply border-0;
        }

        & .cm-scroller {
            @apply text-15 font-mono leading-160;
        }
        & .cm-gutters {
            @apply bg-backgroundColorB;
            @apply border-borderColor7;

            & .cm-lineNumbers .cm-gutterElement {
                @apply text-center text-textColor7;
                @apply px-16;
            }
            & .cm-foldGutter .cm-gutterElement {
                @apply pr-4;
                & span {
                    vertical-align: top;
                    line-height: 13px;
                    &[title="Unfold line"] {
                        line-height: 20px;
                    }
                }
            }
        }
        & .cm-content {
            & .cm-line {
                @apply pl-8 pr-8;
            }
        }
    }
}

/* General overrides */
/* #page-editor { */
/*     & .ace_editor { */
/*         line-height: 170%; */
/*     } */
/*     & .ace_gutter, */
/*     & .ace_gutter-layer { */
/*         width: 70px !important; */

/*         & .ace_gutter-cell { */
/*             padding-left: 0px !important; */
/*             padding-right: 16px !important; */
/*             text-align: center; */
/*         } */
/*     } */
/*     & .ace_scroller { */
/*         left: 70px !important; */
/*     } */
/*     & .ace_scroller.ace_scroll-left { */
/*         box-shadow: 17px 2px 16px -16px rgba(79, 114, 205, 0.3) inset; */
/*     } */

/*     /1* Stop brackets from filling the full line height *1/ */
/*     & .ace_bracket { */
/*         max-height: 20px; */
/*         margin-top: 3px; */
/*     } */
/*     & .ace_dialog.ace_dialog-bottom { */
/*         @apply border-t-1 border-borderColor6; */
/*         @apply bg-backgroundColorB; */
/*         @apply pl-16; */
/*         &, */
/*         & input { */
/*             @apply text-15 font-semibold font-mono; */
/*             @apply text-darkGreen; */
/*         } */
/*         & input { */
/*             @apply w-100; */
/*             background: transparent; */
/*             @apply pl-4; */
/*             &:focus { */
/*                 outline: none !important; */
/*                 border: none !important; */
/*                 box-shadow: none !important; */
/*             } */
/*         } */
/*     } */
/* } */

/* /1* Light theme overrides *1/ */
/* #page-editor .ace-tm { */
/*     & .ace_gutter { */
/*         @apply bg-backgroundColorA; */
/*     } */
/*     & .ace_gutter-cell { */
/*         @apply text-textColor7; */
/*     } */
/*     /1* Same as the content active line *1/ */
/*     & .ace_gutter-cell.ace_gutter-active-line { */
/*         background: rgba(0, 0, 0, 0.07) !important; */
/*     } */
/* } */

/* #page-editor .ace-cobalt { */
/*     & .ace_gutter, */
/*     & .ace_scroller { */
/*         @apply bg-backgroundColorA; */
/*     } */
/* } */

.c-modal.modal-editor-hints {
    & .--modal-content {
        max-width: 1100px;
    }
}
