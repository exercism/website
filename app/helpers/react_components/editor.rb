module ReactComponents
  class Editor < ReactComponent
    initialize_with :solution

    def to_s
      super(
        "editor",
        {
          default_submissions: submissions,
          default_files: SerializeEditorFiles.(solution.files_for_editor),
          default_settings: {
            tab_size: track.indent_size,
            use_soft_tabs: track.indent_style == :space
          },
          autosave: {
            key: solution.uuid,
            save_interval: 500
          },
          panels: {
            instructions: {
              introduction:,
              assignment: SerializeExerciseAssignment.(solution),
              debugging_instructions:
            },
            tests: solution.exercise.practice_exercise? ? {
              tests: solution.tests,
              highlightjs_language: track.highlightjs_language
            } : nil,
            results: {
              test_runner: {
                average_test_duration: track.average_test_duration
              }
            }
          },
          track: {
            title: track.title,
            slug: track.slug
          },
          exercise: {
            title: solution.exercise.title,
            slug: solution.exercise.slug
          },
          links: {
            run_tests: Exercism::Routes.api_solution_submissions_url(solution.uuid),
            back: Exercism::Routes.track_exercise_path(track, solution.exercise)
          }
        }
      )
    end

    private
    def submissions
      submission = solution.latest_submission

      submission ? [SerializeSubmission.(submission)] : []
    end

    memoize
    def introduction
      Markdown::Parse.(solution.introduction)
    end

    memoize
    def instructions
      Markdown::Parse.(solution.instructions)
    end

    memoize
    def debugging_instructions
      return if track.debugging_instructions.blank?

      Markdown::Parse.(track.debugging_instructions)
    end

    memoize
    def track
      solution.track
    end
  end
end
