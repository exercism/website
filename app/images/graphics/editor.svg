<svg
  id="graphic-editor"
  viewBox="0 0 66 66"
  fill="none"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M1 33C1 41.4869 4.37142 49.6263 10.3726 55.6274C16.3737 61.6286 24.5131 65 33 65C41.4869 65 49.6263 61.6286 55.6274 55.6274C61.6286 49.6263 65 41.4869 65 33C65 24.5131 61.6286 16.3737 55.6274 10.3726C49.6263 4.37142 41.4869 1 33 1C24.5131 1 16.3737 4.37142 10.3726 10.3726C4.37142 16.3737 1 24.5131 1 33Z"
    fill="#E8F4FA"
    stroke="#DAEDF7"
    stroke-width="1.4"
    stroke-linecap="round"
    stroke-linejoin="round"
  />
  <path
    opacity="0.15"
    d="M10.1978 50.9464C10.1978 51.4478 12.5872 51.9286 16.8406 52.2831C21.0939 52.6377 26.8626 52.8368 32.8778 52.8368C38.8929 52.8368 44.6616 52.6377 48.9149 52.2831C53.1683 51.9286 55.5578 51.4478 55.5578 50.9464C55.5578 50.4451 53.1683 49.9642 48.9149 49.6097C44.6616 49.2552 38.8929 49.056 32.8778 49.056C26.8626 49.056 21.0939 49.2552 16.8406 49.6097C12.5872 49.9642 10.1978 50.4451 10.1978 50.9464Z"
    fill="#45413C"
  />
  <path
    d="M58.5375 43.1711H7.21748C7.00531 43.1711 6.80183 43.2554 6.6518 43.4055C6.50177 43.5555 6.41748 43.759 6.41748 43.9711V47.1903H59.3375V43.9711C59.3375 43.759 59.2532 43.5555 59.1032 43.4055C58.9531 43.2554 58.7496 43.1711 58.5375 43.1711Z"
    fill="#F0F0F0"
    stroke="#45413C"
    stroke-width="1.4"
    stroke-linecap="round"
    stroke-linejoin="round"
  />
  <path
    d="M59.3375 47.1904H6.41748C6.8203 47.9957 7.43936 48.6729 8.20531 49.1463C8.97127 49.6196 9.85387 49.8704 10.7543 49.8704H55.0223C55.916 49.87 56.7921 49.621 57.5524 49.1512C58.3128 48.6815 58.9275 48.0095 59.3279 47.2104L59.3375 47.1904Z"
    fill="#BDBEC0"
    stroke="#45413C"
    stroke-width="1.4"
    stroke-linecap="round"
    stroke-linejoin="round"
  />
  <path
    d="M53.7199 15.0369H12.0399C11.6156 15.0369 11.2086 15.2054 10.9086 15.5055C10.6085 15.8056 10.4399 16.2125 10.4399 16.6369V43.1713H55.3199V16.6369C55.3199 16.2125 55.1514 15.8056 54.8513 15.5055C54.5513 15.2054 54.1443 15.0369 53.7199 15.0369Z"
    fill="#00B8F0"
  />
  <path
    d="M53.7199 15.0369H12.0399C11.6156 15.0369 11.2086 15.2054 10.9086 15.5055C10.6085 15.8056 10.4399 16.2125 10.4399 16.6369V43.1713H55.3199V16.6369C55.3199 16.2125 55.1514 15.8056 54.8513 15.5055C54.5513 15.2054 54.1443 15.0369 53.7199 15.0369Z"
    stroke="#45413C"
    stroke-width="1.4"
    stroke-miterlimit="10"
  />
  <path
    d="M32.8773 38.3736C30.9904 38.3786 29.1212 38.0094 27.3779 37.2872C25.6346 36.5651 24.0519 35.5043 22.7213 34.1664L22.2253 33.6704C21.9469 33.392 21.7905 33.0145 21.7905 32.6208C21.7905 32.2271 21.9469 31.8496 22.2253 31.5712C22.5037 31.2928 22.8812 31.1364 23.2749 31.1364C23.6686 31.1364 24.0461 31.2928 24.3245 31.5712L24.8205 32.0672C25.8786 33.1253 27.1347 33.9646 28.5171 34.5373C29.8996 35.1099 31.3813 35.4047 32.8777 35.4047C34.3741 35.4047 35.8558 35.1099 37.2382 34.5373C38.6207 33.9646 39.8768 33.1253 40.9349 32.0672L41.4293 31.5712C41.7077 31.2926 42.0853 31.136 42.4791 31.1359C42.8729 31.1357 43.2507 31.292 43.5293 31.5704C43.8079 31.8488 43.9645 32.2264 43.9646 32.6202C43.9648 33.014 43.8085 33.3918 43.5301 33.6704L43.0349 34.1664C41.7039 35.5042 40.1209 36.5649 38.3773 37.287C36.6338 38.0092 34.7644 38.3785 32.8773 38.3736Z"
    fill="white"
    stroke="#009FD9"
    stroke-width="1.4"
    stroke-miterlimit="10"
  />
  <path
    d="M27.3584 29.6999C26.9648 29.6995 26.5875 29.5429 26.3093 29.2645C26.031 28.9862 25.8746 28.6087 25.8744 28.2151V24.0191C25.8698 23.8213 25.9048 23.6246 25.9774 23.4404C26.0499 23.2563 26.1585 23.0886 26.2968 22.947C26.4351 22.8055 26.6003 22.693 26.7827 22.6162C26.965 22.5394 27.1609 22.4999 27.3588 22.4999C27.5567 22.4999 27.7526 22.5394 27.935 22.6162C28.1174 22.693 28.2826 22.8055 28.4208 22.947C28.5591 23.0886 28.6677 23.2563 28.7403 23.4404C28.8128 23.6246 28.8478 23.8213 28.8432 24.0191V28.2151C28.843 28.6089 28.6865 28.9864 28.4081 29.2648C28.1297 29.5432 27.7522 29.6997 27.3584 29.6999Z"
    fill="white"
    stroke="#009FD9"
    stroke-width="1.4"
    stroke-miterlimit="10"
  />
  <path
    d="M38.3969 29.7C38.0032 29.6996 37.6258 29.543 37.3475 29.2647C37.0691 28.9863 36.9125 28.6089 36.9121 28.2152V24.0192C36.9121 23.8242 36.9505 23.6312 37.0251 23.451C37.0998 23.2709 37.2091 23.1072 37.347 22.9693C37.4849 22.8314 37.6486 22.7221 37.8287 22.6474C38.0088 22.5728 38.2019 22.5344 38.3969 22.5344C38.5919 22.5344 38.785 22.5728 38.9651 22.6474C39.1453 22.7221 39.3089 22.8314 39.4468 22.9693C39.5847 23.1072 39.6941 23.2709 39.7687 23.451C39.8433 23.6312 39.8817 23.8242 39.8817 24.0192V28.2152C39.8813 28.6089 39.7247 28.9863 39.4464 29.2647C39.168 29.543 38.7906 29.6996 38.3969 29.7Z"
    fill="white"
    stroke="#009FD9"
    stroke-width="1.4"
    stroke-miterlimit="10"
  />
</svg>
