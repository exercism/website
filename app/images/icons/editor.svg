<svg id="editor" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
  <defs>
    <style>
      .a {
        fill:none;
        stroke:currentColor;
        stroke-linecap:round;
        stroke-linejoin:round;
        stroke-width:1.5px;
      }
    </style>
  </defs>
  <title>file-code-1</title>
  <path class="a" d="M22.5,21.753a1.5,1.5,0,0,1-1.5,1.5H3a1.5,1.5,0,0,1-1.5-1.5V2.253A1.5,1.5,0,0,1,3,.753H18.045a1.5,1.5,0,0,1,1.048.427l2.955,2.882A1.5,1.5,0,0,1,22.5,5.136Z"/>
  <path class="a" d="M10.5,8.253H9.75a1.409,1.409,0,0,0-1.5,1.294v1.247a1.256,1.256,0,0,1-.668,1.076L6,12.78l1.582.909a1.258,1.258,0,0,1,.668,1.076v1.248a1.408,1.408,0,0,0,1.5,1.293h.75"/>
  <path class="a" d="M13.5,8.253h.75a1.408,1.408,0,0,1,1.5,1.294v1.247a1.256,1.256,0,0,0,.668,1.076L18,12.78l-1.582.909a1.258,1.258,0,0,0-.668,1.076v1.248a1.408,1.408,0,0,1-1.5,1.293H13.5"/>
</svg>