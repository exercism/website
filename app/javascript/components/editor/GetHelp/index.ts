export { GetHelpTab } from './GetHelpTab'
export { GetHelpPanel } from './GetHelpPanel'
