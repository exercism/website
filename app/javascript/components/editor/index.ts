export { TestRun, TestRunStatus } from './types'
export { Props, EditorFeatures as EditorFeatures, TaskContext } from './Props'
export { Header } from './Header'
export { FileEditorCodeMirror, FileEditorHandle } from './FileEditorCodeMirror'
export { InstructionsPanel } from './InstructionsPanel'
export { TestsPanel } from './TestsPanel'
export { ResultsPanel } from './ResultsPanel'
export { InstructionsTab } from './InstructionsTab'
export { TestsTab } from './TestsTab'
export { ResultsTab } from './ResultsTab'
export { EditorStatusSummary } from './EditorStatusSummary'
export { RunTestsButton } from './RunTestsButton'
export { SubmitButton } from './SubmitButton'
export { useSaveFiles } from './useSaveFiles'
export { useEditorFiles } from './useEditorFiles'
export { useEditorFocus } from './useEditorFocus'
export { useSubmissionsList } from './useSubmissionsList'
export { useFileRevert } from './useFileRevert'
export { useIteration } from './useIteration'
export { useDefaultSettings } from './useDefaultSettings'
export { useEditorStatus, EditorStatus } from './useEditorStatus'
export { useEditorTestRunStatus } from './useEditorTestRunStatus'
export { useSubmissionCancelling } from './useSubmissionCancelling'
