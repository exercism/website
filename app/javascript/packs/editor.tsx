import React from 'react'
import { initReact } from '../utils/react-bootloader.jsx'

import '../../css/pages/editor'
import '../../css/modals/editor-hints'

import { Editor, Props } from '../components/Editor'

import { camelizeKeys } from 'humps'
function camelizeKeysAs<T>(object: any): T {
  return (camelizeKeys(object) as unknown) as T
}

// Add all react components here.
// Each should map 1-1 to a component in app/helpers/components
initReact({
  editor: (data: any) => <Editor {...camelizeKeysAs<Props>(data)} />,
})
