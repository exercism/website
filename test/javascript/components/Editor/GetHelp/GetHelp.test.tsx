import React from 'react'
import { render, RenderResult, screen } from '@testing-library/react'
import { GetHelpPanel } from '@/components/editor/GetHelp'
import { MOCK_DATA } from './mockdata'

describe('GetHelp tests', () => {
  let renderedComponent: RenderResult

  beforeEach(() => {
    renderedComponent = render(
      <GetHelpPanel
        assignment={MOCK_DATA.ASSIGNMENT}
        helpHtml={MOCK_DATA.HELP}
      />
    )
  })

  test('Component snapshot test', () => {
    const { asFragment } = renderedComponent

    expect(asFragment()).toMatchSnapshot()
  })

  test('Component renders both hints and help', () => {
    const hint = screen.getByText(/Define the expected oven time in minutes/)
    const help = screen.getByText(/To get help if you're having trouble/)

    expect(hint).toBeInTheDocument()
    expect(help).toBeInTheDocument()
  })
})
