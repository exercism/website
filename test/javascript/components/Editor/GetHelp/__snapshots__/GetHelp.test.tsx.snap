// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetHelpPanel tests Component snapshot test 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="tab-get-help"
    class="--tab-panel undefined"
    id="panel-get-help"
    role="tabpanel"
    style="display: none;"
    tabindex="0"
  >
    <div
      class="px-24 flex flex-col gap-24"
    >
      <div
        class="flex flex-col"
      >
        <header
          class="flex items-center gap-12 mb-24\\""
        >
          <img
            alt=""
            class="c-icon"
            role="presentation"
            src="undefined/assets/undefined"
          />
          <h2>
            Hints and Tips
          </h2>
        </header>
        <details
          class="c-details"
        >
          <summary
            class="--summary"
          >
            <div
              class="--summary-inner"
            >
              1. Define the expected oven time in minutes
              <img
                alt=""
                class="c-icon --closed-icon"
                role="presentation"
                src="undefined/assets/undefined"
              />
              <img
                alt=""
                class="c-icon --open-icon"
                role="presentation"
                src="undefined/assets/undefined"
              />
            </div>
          </summary>
          <div
            class="c-textual-content --small"
          >
            <ul>
              

              <li>
                You need to define a 
                <a
                  href="https://www.rubyguides.com/2017/07/ruby-constants/"
                  rel="noreferrer"
                  target="_blank"
                >
                  constant
                </a>
                 that should contain the

                <a
                  href="https://ruby-doc.org/core-2.7.0/Integer.html"
                  rel="noreferrer"
                  target="_blank"
                >
                  integer
                </a>
                 value specified in the recipe.
              </li>
              

              <li>
                The 
                <code>
                  ::
                </code>
                 used in 
                <code>
                  Lasagna::EXPECTED_MINUTES_IN_OVEN
                </code>
                 means that 
                <code>
                  EXPECTED_MINUTES_IN_OVEN
                </code>
                 needs to be defined
within the 
                <code>
                  Lasagna
                </code>
                 class.
              </li>
              

            </ul>
            

          </div>
        </details>
        <details
          class="c-details"
        >
          <summary
            class="--summary"
          >
            <div
              class="--summary-inner"
            >
              2. Calculate the remaining oven time in minutes
              <img
                alt=""
                class="c-icon --closed-icon"
                role="presentation"
                src="undefined/assets/undefined"
              />
              <img
                alt=""
                class="c-icon --open-icon"
                role="presentation"
                src="undefined/assets/undefined"
              />
            </div>
          </summary>
          <div
            class="c-textual-content --small"
          >
            <ul>
              

              <li>
                You need to define a 
                <a
                  href="https://launchschool.com/books/ruby/read/methods"
                  rel="noreferrer"
                  target="_blank"
                >
                  method
                </a>
                 with a single parameter for the
actual time so far.
              </li>
              

              <li>
                You can 
                <a
                  href="https://www.freecodecamp.org/news/idiomatic-ruby-writing-beautiful-code-6845c830c664/#implicit-return"
                  rel="noreferrer"
                  target="_blank"
                >
                  implicitly return an
integer
                </a>
                 from
the method.
              </li>
              

              <li>
                You can use the 
                <a
                  href="https://www.w3resource.com/ruby/ruby-arithmetic-operators.php"
                  rel="noreferrer"
                  target="_blank"
                >
                  mathematical operator for
subtraction
                </a>
                 to subtract values.
              </li>
              

            </ul>
            

          </div>
        </details>
        <details
          class="c-details"
        >
          <summary
            class="--summary"
          >
            <div
              class="--summary-inner"
            >
              3. Calculate the preparation time in minutes
              <img
                alt=""
                class="c-icon --closed-icon"
                role="presentation"
                src="undefined/assets/undefined"
              />
              <img
                alt=""
                class="c-icon --open-icon"
                role="presentation"
                src="undefined/assets/undefined"
              />
            </div>
          </summary>
          <div
            class="c-textual-content --small"
          >
            <ul>
              

              <li>
                You need to define a 
                <a
                  href="https://launchschool.com/books/ruby/read/methods"
                  rel="noreferrer"
                  target="_blank"
                >
                  method
                </a>
                 with a single parameter for the
number of layers.
              </li>
              

              <li>
                You can 
                <a
                  href="https://www.freecodecamp.org/news/idiomatic-ruby-writing-beautiful-code-6845c830c664/#implicit-return"
                  rel="noreferrer"
                  target="_blank"
                >
                  implicitly return an
integer
                </a>
                 from
the method.
              </li>
              

              <li>
                You can use the 
                <a
                  href="https://www.w3resource.com/ruby/ruby-arithmetic-operators.php"
                  rel="noreferrer"
                  target="_blank"
                >
                  mathematical operator for
multiplication
                </a>
                 to multiply values.
              </li>
              

              <li>
                You could define an extra constant for the time in minutes per layer, or use a "magic number" in the code.
              </li>
              

            </ul>
            

          </div>
        </details>
        <details
          class="c-details"
        >
          <summary
            class="--summary"
          >
            <div
              class="--summary-inner"
            >
              4. Calculate the total working time in minutes
              <img
                alt=""
                class="c-icon --closed-icon"
                role="presentation"
                src="undefined/assets/undefined"
              />
              <img
                alt=""
                class="c-icon --open-icon"
                role="presentation"
                src="undefined/assets/undefined"
              />
            </div>
          </summary>
          <div
            class="c-textual-content --small"
          >
            <ul>
              

              <li>
                You need to define a 
                <a
                  href="https://launchschool.com/books/ruby/read/methods"
                  rel="noreferrer"
                  target="_blank"
                >
                  method
                </a>
                 with two named parameters:

                <code>
                  number_of_layers
                </code>
                 and 
                <code>
                  actual_minutes_in_oven
                </code>
                .
              </li>
              

              <li>
                You can 
                <a
                  href="https://www.freecodecamp.org/news/idiomatic-ruby-writing-beautiful-code-6845c830c664/#implicit-return"
                  rel="noreferrer"
                  target="_blank"
                >
                  implicitly return an
integer
                </a>
                 from
the method.
              </li>
              

              <li>
                You can 
                <a
                  href="http://ruby-for-beginners.rubymonstas.org/objects/calling.html"
                  rel="noreferrer"
                  target="_blank"
                >
                  invoke
                </a>
                 one of the other methods you've
defined previously.
              </li>
              

              <li>
                You can use the 
                <a
                  href="https://www.w3resource.com/ruby/ruby-arithmetic-operators.php"
                  rel="noreferrer"
                  target="_blank"
                >
                  mathematical operator for addition
                </a>
                
to add values.
              </li>
              

            </ul>
            

          </div>
        </details>
      </div>
      <div
        class="flex flex-col"
      >
        <header
          class="flex items-center gap-12 mb-24"
        >
          <img
            alt=""
            class="c-icon"
            height="40"
            role="presentation"
            src="undefined/assets/undefined"
            width="40"
          />
          <h2
            class="text-h3"
          >
            Help
          </h2>
        </header>
        <div
          class="c-textual-content --base"
        >
          <p>
            To get help if you're having trouble, you can use one of the following resources:
          </p>
          
  
          <ul>
            
  
            <li>
              <a
                href="http://ruby-doc.org/"
                rel="noreferrer"
                target="_blank"
              >
                Ruby Documentation
              </a>
            </li>
            
  
            <li>
              <a
                href="http://stackoverflow.com/questions/tagged/ruby"
                rel="noreferrer"
                target="_blank"
              >
                StackOverflow
              </a>
            </li>
            
  
            <li>
              
  
              <a
                href="https://www.reddit.com/r/ruby"
                rel="noreferrer"
                target="_blank"
              >
                /r/ruby
              </a>
               is the Ruby subreddit.
            </li>
            
  
          </ul>
          
  
          <p>
            Should those resources not suffice, check the following pages:
          </p>
          
  
          <ul>
            
  
            <li>
              The 
              <a
                href="https://exercism.org/docs/tracks/ruby"
              >
                Ruby track's documentation
              </a>
              
  
            </li>
            
  
            <li>
              The 
              <a
                href="https://forum.exercism.org/c/programming/ruby"
                rel="noreferrer"
                target="_blank"
              >
                Ruby track's programming category on the forum
              </a>
              
  
            </li>
            
  
            <li>
              <a
                href="https://forum.exercism.org/c/programming/5"
                rel="noreferrer"
                target="_blank"
              >
                Exercism's programming category on the forum
              </a>
            </li>
            
  
            <li>
              The 
              <a
                href="https://exercism.org/docs/using/faqs"
              >
                Frequently Asked Questions
              </a>
              
  
            </li>
            
  
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
